### DISCUSSION TITLE
Validation of Complex Properties currently not supported by Stylet

### DISCUSSION THREAD

**Discussion by MrCoffeeAtGitHub** [canton7/Stylet#285] [Ideas]
I've found this [old thread](https://github.com/canton7/Stylet/discussions/236), but I'd like to open a new one even it is simular issue

I changed the Stylet-Example of Validating to reproduce the issue in my own fork (https://github.com/MrCoffeeAtGitHub/Stylet/tree/ComplexPropertyValidation) 

Mainly, I moved from the properties out of ViewModel into an own Model and using this. 

```cs
public class UserModel : PropertyChangedBase
    {
        private string _userName;
        private string _email;
        private string _password;
        private string _passwordConfirmation;
        private Stringable<int> _age;

        public string UserName
        {
            get => _userName;
            set => SetAndNotify(ref _userName, value);
        }

        public string Email
        {
            get => _email;
            set => SetAndNotify(ref _email, value);
        }

        public string Password
        {
            get => _password; set => SetAndNotify(ref _password, value);
        }
        public string PasswordConfirmation
        {
            get => _passwordConfirmation;
            set => SetAndNotify(ref _passwordConfirmation, value);
        }
        public Stringable<int> Age
        {
            get => _age;
            set => SetAndNotify(ref _age, value);
        }
```

At the ViewModel, I instanciate that and provide it a property for this:

```cs   
private UserModel _userModel = new UserModel();
        public UserModel UserModel
        {
            get => _userModel;
            set => SetAndNotify(ref _userModel, value);
        }
```
The validation itself is now in two classes so we can reuse the code of UserModelValidator and IoC instanciates it correctly.

```cs
public class UserViewModelValidator : AbstractValidator<UserViewModel>
  {
        public UserViewModelValidator(UserModelValidator modelValidator)
        {
            RuleFor(x => x.UserModel).SetValidator(modelValidator);
        }
    }

    public class UserModelValidator : AbstractValidator<UserModel>
    {
        public UserModelValidator()
        {
            RuleFor(x => x.UserName).NotEmpty().Length(1, 20);
            RuleFor(x => x.Email).NotEmpty().EmailAddress();
            RuleFor(x => x.Password).NotEmpty().Matches("[0-9]").WithMessage("{PropertyName} must contain a number");
            RuleFor(x => x.PasswordConfirmation).Equal(s => s.Password).WithMessage("{PropertyName} should match Password");
            RuleFor(x => x.Age).Must(x => x.IsValid).WithMessage("{PropertyName} must be a valid number");
            When(x => x.Age.IsValid, () =>
            {
                RuleFor(x => x.Age.Value).GreaterThan(0).WithName("Age");
            });
        }
    }
```

The validation runs and finds all errors so submit is blocked but the fields aren't marked as error (red) anymore. The interface INotifyDataErrorInfo is not implemented in PropertyBase so, I'd need to implement it by myself but I think this would great if Stylet could do this in PropertyBase, so all properties could be checked. This has been already discussed on [Stackoverflow](https://stackoverflow.com/questions/20012753/idataerrorinfo-does-not-fire-when-binding-to-a-complex-object) and [Social MSDN](https://social.msdn.microsoft.com/Forums/vstudio/en-US/f8cc491c-8975-4494-906b-593970dcaeca/idataerrorinfo-and-long-binding-path?forum=wpf).

What is your opinion about this?

**Comment by canton7** [comment]
So your problem is that you extend `PropertyChangedBase`, but that doesn't implement `INotifyDataErrorInfo`? Why not extend `ValidatingModelBase` or `Screen`? Then your `UserViewModel` will be able to validate itself.
